### ISSUE TITLE
Kernel tries to recompile when building system from nix-shell

### ISSUE THREAD

**Issue by Synthetica9** [NixOS/nixpkgs#51681]
## Issue description

When building a recent nixpkgs version kernel with `nixos-rebuild`, `nix build`, and probably others, nix wants to recompile the kernel when we are currently in a `nix-shell`. 

This seems to be related to the fact that these two packages get added to the derivation of the kernel:
- `ncurses-6.1-20181027`
- `pkg-config-0.29.2`

[Full diff](https://gist.github.com/Synthetica9/c30cc08d3b997b02fd36f10e34f93027)

This doesn't happen for older kernel/nixpkgs combinations, specifically linux-4.14.76, nixpkgs rev 81f5c2698a8 (2018 Oct 17). 

cc @tilpner, who was also able to reproduce this on irc. 

### Steps to reproduce

Minimal failing example:

```sh
# Doesn't try to build kernel:
nixos-rebuild build

# Tries to rebuild kernel:
nix-shell -p [] --run "nixos-rebuild build"
```

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.85, NixOS, 19.03pre161818.776f084cf1b (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.1.3`
 - channels(root): `"nixos-19.03pre161797.c26dbef830a"`
 - channels(synthetica): `"nixos-19.03pre158246.6141939d6e0"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by Synthetica9** [comment]
Yes! Thank you! Reverting that exact (merge) commit in my local repo gives the expected behaviour:

```shell
$ nix-shell -p [] --run "nix build -f nixos system"
trace: Optimising i3-gaps
trace: Optimising ag
trace: Optimising htop
trace: Optimising netcat-gnu
[1/13/23 built, 0.0 MiB DL]
error: interrupted by the user
$ # Building kernel
$ git revert -m 1 c67ccd34c08fdb4e97c647c3fb09c1eb6e5431c6
[detached HEAD 9a59ea57b52] Revert "Merge pull request #47909 from teto/menuconfig"
 1 file changed, 2 deletions(-)
$ nix-shell -p [] --run "nix build -f nixos system"
trace: Optimising i3-gaps
trace: Optimising ag
trace: Optimising htop
trace: Optimising netcat-gnu
[18 built, 0.0 MiB DL
$ # Not building kernel! 
```

**Comment by tilpner** [comment]
nixos-rebuild could unset IN_NIX_SHELL

**Comment by Synthetica9** [comment]
I'ts not just nixos-rebuild, it's also `nix-build`, `nix build`, `nix-instantiate`, and probably more

**Comment by FRidh** [comment]
cc @teto @joachifm @yegortimoshenko 

`nix-shell` is for development and debugging of a package, and should therefore resemble the actual build as close as possible. Automatically bringing in additional dependencies when developing is perhaps not such a good idea. Can't  simply `-p` be used to bring in `pkgconfig` and `ncurses`?

**Comment by yegortimoshenko** [comment]
Just had this same problem.

I think this use case warrants adding `shellInputs` attribute handling to `nix-shell`.

**Comment by matthewbauer** [comment]
You can kind of get this with shellHook:

```
shellHook = lib.concatMapStringsSep "\n" (pkg: "activatePackage ${pkg}") [ pkgconfig ncurses ];
```

It should only evaluate when running nix-shell on the derivation.

**Comment by teto** [comment]
This is a consequence I realized was possible only after submitting my PR and it was merged by then already. As a short term fix, reverting may be the way to go. As someone who works quite often on the kernel source, or on packages, I believe default shellHooks could definitely be improved for a wide range of applications (https://github.com/NixOS/nixpkgs/pull/42864). So if we had a way to modify shellHooks that play nice with the rest of nix, that would be great.
